src/cmap/commands.ts

```diff
@@ -709,3 +715,54 @@ export class BinMsg {
     return { utf8: { writeErrors: false } };
   }
 }
+
+const MESSAGE_HEADER_SIZE = 16;
+const COMPRESSION_DETAILS_SIZE = 9; // originalOpcode + uncompressedSize, compressorID
+
+/**
+ * @internal
+ *
+ * An OP_COMPRESSED request wraps either an OP_QUERY or OP_MSG message.
+ */
+export class OpCompressedRequest {
+  constructor(
+    private command: WriteProtocolMessageType,
+    private options: { zlibCompressionLevel: number; agreedCompressor: CompressorName }
+  ) {}
+
+  // Return whether a command contains an uncompressible command term
+  // Will return true if command contains no uncompressible command terms
+  static canCompress(command: WriteProtocolMessageType) {
+    const commandDoc = command instanceof OpMsgRequest ? command.command : command.query;
+    const commandName = Object.keys(commandDoc)[0];
+    return !uncompressibleCommands.has(commandName);
+  }
+
+  async toBin(): Promise<Buffer[]> {
+    const concatenatedOriginalCommandBuffer = Buffer.concat(this.command.toBin());
```

Comment:
Is it necessary to concat here? Doesn't toBin return the chunks already sliced appropriately?

Same concern for `compressCommand`

It is not for this PR to fix but I think we identified that we should be writing chunks to the stream rather than reallocing and copying bytes. I would hope the code we introduce could facilitate that improvement.

Reply:
No, `toBin()` currently returns an array of buffers.  There's no change in the existing serialization behavior here.

Is this something you'd want me to fix now?  I can, but I intentionally limited the changes to the connection layer (and some refactors).  The core serialization and compression logic is unchanged

Reply:
Done: https://jira.mongodb.org/browse/NODE-5740